internal/handlers/handlers_test.go

```diff
@@ -0,0 +1,89 @@
+package handlers
+
+import (
+	gcppubsub "cloud.google.com/go/pubsub"
+	"github.com/stretchr/testify/suite"
+	"github.com/thought-machine/spot-interruption-exporter/internal/cache"
+	"github.com/thought-machine/spot-interruption-exporter/internal/handlers/test_data"
+	"github.com/thought-machine/spot-interruption-exporter/internal/metrics/mocks"
+	"go.uber.org/zap"
+	"testing"
+	"time"
+)
+
+type HandlersTestSuite struct {
+	suite.Suite
+	mockMetrics *mocks.Client
+	l           *zap.SugaredLogger
+}
+
+var (
+	mockInterruptionMessage = &gcppubsub.Message{
+		ID:   "12345",
+		Data: test_data.InterruptionEventJSONFile,
+	}
+	mockCreationMessage = &gcppubsub.Message{
+		ID:   "12345",
+		Data: test_data.CreationEventJSONFile,
+	}
+)
+
+func (suite *HandlersTestSuite) SetupSuite() {
+	suite.mockMetrics = mocks.NewClient(suite.T())
+	l, err := zap.NewDevelopment()
+	suite.NoError(err)
+	suite.l = l.Sugar()
+}
+func TestHandlersTestSuite(t *testing.T) {
+	suite.Run(t, new(HandlersTestSuite))
+}
+
+func (suite *HandlersTestSuite) TestHandleInterruptionEvents() {
+	suite.mockMetrics.EXPECT().IncreaseInterruptionEventCounter("fake-cluster").Times(1)
+	initialInstances := map[string]string{
+		"projects/mock-project/zones/europe-west1-c/instances/mock-instance-spot-3706-5b909138-nr65": "fake-cluster",
+	}
+	instanceToClusterMappings := cache.NewCacheWithTTLFrom(cache.NoExpiration, initialInstances)
+	interruptions := make(chan *gcppubsub.Message)
+	go HandleInterruptionEvents(interruptions, instanceToClusterMappings, suite.mockMetrics, suite.l)
```

Comment:
This looks fragile - don't you need something to assert that the messages have been processed before the expectations are checked? Maybe close the channel and add a WaitGroup?

Ditto below

Reply:
What expectations are you referring to? The sender should always close the channel, which it does. I don't know what benefit adding a waitgroup here would give.

Reply:
The expectation on line 42 (which I am assuming is checked once the test function terminates). That expectation could be checked more-or-less instantly after the second message is pushed to the channel, before the handler finishes processing it (which means this test would not necessarily enforce the deduplication behaviour). My suggestion is:

```
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		HandleInterruptionEvents(interruptions, instanceToClusterMappings, suite.mockMetrics, suite.l)
		wg.Done()
	} ()
	interruptions <- mockInterruptionMessage
	interruptions <- mockInterruptionMessage
	close(interruptions)
	wg.Wait()
```

Reply:
Ah sorry I thought this was `main.go`, didn't realise this was the test! This was actually terrible on my part, if I had properly passed the waitgroup as a part of the handler then it actually cleans up this test a lot. I've fixed it now if you want to have another look. Good shout